### ISSUE TITLE
[Design Doc] Stdlib-Only Constraint: Impact and Mitigation Strategy

### ISSUE THREAD

**Issue by h315uk3** [h315uk3/symbiosis#45]
## Purpose
Document the Python stdlib-only constraint, its impact on analysis capabilities, and the mitigation strategy chosen for with-me plugin implementation.

## Constraint Definition

**Statement:** The with-me plugin must use only Python 3.11+ standard library. No external dependencies (NumPy, SciPy, sklearn, matplotlib, etc.).

**Rationale:**
- Project philosophy: Local-first, self-contained, zero external dependencies
- Deployment flexibility: No package installation required
- Simplicity: Minimize complexity, easier to audit and maintain

**Scope:**
- ✅ Allowed: Modules in Python stdlib (math, statistics, collections, json, etc.)
- ✅ Allowed: Custom implementations of algorithms
- ❌ Prohibited: External packages (pip install xyz)

---

### Phase 1: Bayesian Belief Updating ✅ FULLY FEASIBLE

**Required Operations:**
- Shannon entropy: H = -Σ p(h) log₂ p(h)
  - ✅ math.log2() available
  - ✅ Simple sum/iteration sufficient
  
- Bayesian update: p₁(h) ∝ p₀(h) × L(obs|h)
  - ✅ Basic arithmetic only
  - ✅ Dict operations for distribution storage
  
- Likelihood estimation L(answer|h)
  - ✅ Keyword matching (string operations)
  - ✅ Laplace smoothing (basic arithmetic)
  - ✅ No external NLP/ML required

**Feasibility:** ✅ 100% feasible
**Estimate:** 50h total (Phase 1-A, 1-B, 1-C, 1-D)

---

### Phase 2A: Basic Statistical Analysis ✅ FEASIBLE (WITH LIMITATIONS)

**Enabled by stdlib:**
- mean, median, stdev: ✅ statistics module
- Histograms: ✅ collections.Counter
- Value aggregation: ✅ sum(), min(), max()
- Percentiles: ⚠️ Manual calculation required

**Example - mean() usage:**
```python
import statistics
values = [1.1, 2.3, 1.8, 0.9]
mean = statistics.mean(values)
median = statistics.median(values)
stdev = statistics.stdev(values)  # Requires N > 1
```

**Disabled (Requires External Libs):**
- Scipy.stats functions (t-test, etc.): ❌ Would need scipy
- Matplotlib graphing: ❌ Would need matplotlib
- Pandas DataFrames: ❌ Would need pandas

**Mitigation:**
- Generate text-based reports (ASCII tables) instead of graphs
- Implement Pearson correlation manually (feasible for N < 100)
- Use distributions empirically (frequency counts, not theoretical distributions)

**Feasibility:** ✅ ~80% feasible (text-based, no hypothesis testing)
**Estimate:** 8h (Phase 2-A)

---

### Phase 2B: Parameter Optimization via Grid Search ✅ FEASIBLE

**Grid Search Advantages:**
- Brute-force search: ✅ Feasible for small parameter spaces
- No derivative computation: ✅ Avoid scipy.optimize
- Evaluation metric: ✅ Average information gain (just aggregation)

**Example - Grid Search pattern:**
```python
results = []
for log_base in [1.5, 2.0, 2.5, 3.0, ..., 10.0]:
    # Evaluate each parameter value
    efficiency = evaluate_sessions_with_log_base(sessions, log_base)
    results.append((log_base, efficiency))

best_log_base = max(results, key=lambda x: x[1])[0]
```

**Disabled (Requires External Libs):**
- Scipy.optimize functions (minimize, differential_evolution): ❌ Would need scipy
- Bayesian optimization: ❌ Would need external package
- Gradient-based optimization: ❌ Would need automatic differentiation

**Mitigation:**
- Limit grid range (1.5-10.0, step 0.5 = 18 points)
- Evaluation time acceptable for ~10-20 session records
- Sensitivity analysis via simple tabulation

**Feasibility:** ✅ 100% feasible (Grid Search is pure logic)
**Estimate:** 4h (Phase 2-B)

---

### Phase 3: Advanced Statistical Analysis ❌ NOT FEASIBLE (DELETED)

**Original Phase 3 Intent:**
- Correlation analysis: Pearson/Spearman correlations between metrics
- Hypothesis testing: t-tests, chi-square tests
- Regression: Linear/logistic regression for parameter sensitivity
- Visualization: Plots, heatmaps, distribution curves

**Disabled by Constraint:**
- Scipy.stats required for: t-test, chi-square, Shapiro-Wilk (normality test)
- Sklearn required for: regression models, cross-validation
- Matplotlib required for: plotting

**Mitigation:**
- ❌ DELETED Phase 3 entirely
- ✅ Keep Phase 2A (basic stats)
- ✅ Keep Phase 2B (Grid Search)
- Defer advanced analysis to future (requires external dependency decision)

**Feasibility:** ❌ 0% feasible
**Decision:** Cancel Phase 3

---

### Available Functions
| Function | Capability | Example |
|----------|-----------|---------|
| mean() | Average | statistics.mean([1, 2, 3]) → 2.0 |
| median() | Middle value | statistics.median([1, 2, 3]) → 2 |
| stdev() | Standard deviation | statistics.stdev([1, 2, 3]) → 1.0 |
| variance() | Variance | statistics.variance([1, 2, 3]) → 1.0 |
| quantiles() | Percentile points | statistics.quantiles([1..10], n=4) → quartiles |

### What's Missing
- No hypothesis testing (requires probability distributions)
- No distribution fitting
- No multivariate analysis
- No confidence intervals (would need t-distribution)

---

## Custom Implementation: Pearson Correlation

**When feasible:** N < ~100 (manual computation acceptable)

**Implementation (manual, no numpy):**
```python
def pearson_correlation(x, y):
    """Compute Pearson r without numpy."""
    if len(x) != len(y) or len(x) < 2:
        return 0.0
    
    x_mean = statistics.mean(x)
    y_mean = statistics.mean(y)
    
    numerator = sum((x[i] - x_mean) * (y[i] - y_mean) 
                    for i in range(len(x)))
    
    x_var = sum((xi - x_mean) ** 2 for xi in x)
    y_var = sum((yi - y_mean) ** 2 for yi in y)
    denominator = (x_var * y_var) ** 0.5
    
    return numerator / denominator if denominator > 0 else 0.0
```

**Limitations:**
- O(N) time complexity (acceptable for N < 100)
- No numerical stability optimization (acceptable for small N)
- Single-threaded (acceptable for small N)

---

## Text-Based Reporting Strategy

Since matplotlib prohibited, use ASCII/text output:

**Option 1: Simple Tables**
```
log_base | Avg Efficiency | Change from Min
---------|----------------|----------------
  1.5    |     0.8125     |    +0.0000
  2.0    |     0.8234     |    +0.0109
  2.5    |     0.8342     |    +0.0217
  ...
```

**Option 2: ASCII Histograms**
```
Efficiency Distribution (5 bins):
[0.0-0.2):  #### (4 questions)
[0.2-0.4):  ######## (8 questions)
[0.4-0.6):  ############ (12 questions)
[0.6-0.8):  ##### (5 questions)
[0.8-1.0]:  ## (2 questions)
```

**Option 3: Summary Reports**
```
=== Session Efficiency Summary ===
Total Questions: 31
Total Information Gained: 18.34 bits
Avg per Question: 0.59 bits
Median: 0.52 bits
Stdev: 0.24 bits
```

---

## Decision Record

| Aspect | Decision | Rationale |
|--------|----------|-----------|
| **Phase 1** | ✅ Proceed | Fully feasible, core functionality |
| **Phase 2A** | ✅ Proceed (Modified) | Feasible with text-based output |
| **Phase 2B** | ✅ Proceed | Grid Search is pure logic, stdlib-only |
| **Phase 3** | ❌ Cancel | Requires scipy.stats for hypothesis testing |
| **Visualization** | Text-based | ASCII tables/histograms instead of graphs |
| **Correlation** | Manual impl. | Pearson formula by hand for N < 100 |

---

## Timeline and Estimate

- Phase 1: 50h (Bayesian core)
- Phase 2A: 8h (Basic statistics)
- Phase 2B: 4h (Grid Search tuning)
- **Total: 62h** (down from original 70h Phase 3 plan)

---

## Future Flexibility

**If external dependencies approved:**
1. NumPy: Matrix operations, advanced statistical functions
2. SciPy: Hypothesis testing, optimization algorithms
3. Matplotlib: Professional visualization
4. Scikit-learn: Machine learning models, regression

**Recommendation:** Keep constraint for Phase 1-2. Revisit in Phase 3 if rigorous statistical analysis justified.

---

## Related Issues
- #37-42: Implementation phases (all feasible)
- #43: Theoretical foundation (Bayesian approach)
- #44: Reward function refinement (depends on Phase 1)

**Comment by h315uk3** [comment]
This design doc is **superseded by actual implementation decisions**.

## What Happened

### Phase 1: Bayesian Belief Updating ✅
- Fully implemented in PR #76
- All stdlib-only as designed
- Modules: `dimension_belief.py` with Shannon entropy, Bayesian update

### Phase 2A/2B: Statistical Analysis & Grid Search ❌
- Initially implemented (commits 9bc15a8, etc.)
- `question_stats.py`, parameter tuning features
- Later removed in Issue #77 cleanup (commit 3557306)
- Decision: Computational skills removed, moved computation to CLI

### Phase 3: Advanced Analysis ❌
- Cancelled as designed (requires scipy.stats)
- Never implemented

## Current Architecture

The stdlib-only constraint **remains in effect**, but the architecture evolved:

**What's Implemented (stdlib-only):**
- ✅ Shannon entropy calculation: `math.log2()`
- ✅ Bayesian update: Basic arithmetic
- ✅ Information gain: Entropy delta
- ✅ Session persistence: `json` module
- ✅ Likelihood validation: `math` operations

**What Was Removed:**
- ❌ Statistical analysis (question_stats.py)
- ❌ Grid search parameter tuning
- ❌ Correlation analysis
- ❌ Text-based reporting

## Architectural Shift

Original design: Python + Skills for all computation

Actual implementation (Issue #77):
- **Python CLI**: All mathematical computation (`with_me/cli/session.py`)
- **LLM**: Semantic understanding only (likelihood estimation)
- **No computational skills**: Removed to solve conversation turn issues

This change maintained the stdlib-only constraint while improving architecture.

## Closing Reason

This design doc served its purpose during Phase 1-2 planning. The constraint analysis was valuable and **the stdlib-only constraint is still respected**. However, the specific implementation approach (stats/grid-search via skills) was superseded by architectural decisions in Issue #77.

**Status**: Design validated ✓, Implementation evolved ✓, Constraint maintained ✓

The stdlib-only constraint remains a core principle of the project and is successfully implemented in the current architecture.
